File: web/src/core/engine.ts
import { DEFAULT_CONFIG, OUTCOMES, type Outcome } from "./constants";
import {
  lambdasFromElo,
  marginalHda,
  scorelineMatrix,
  topScorelines,
  validOdds,
} from "./dixonColes";
import { devig } from "./devig";
import { filterLegs } from "./gate";
import { jointMatchStakes, type Leg } from "./kelly";
import { conditionalScorelineMatrix, type LiveState } from "./live";
import { getRating } from "./ratings";

export {
  lambdasFromElo,
  logFactorial,
  marginalHda,
  poissonPmf,
  scorelineMatrix,
  tau,
  topScorelines,
  validOdds,
} from "./dixonColes";

export interface OutcomeView {
  outcome: Outcome;
  modelP: number;
  fairOdds: number;
  marketPrice: number | null;
  marketOdds: number | null;
  consensusP: number | null;
  edge: number | null;
  evPerDollar: number | null;
  stakeFraction: number;
  stakeUsd: number;
  bet: boolean;
}

export interface MatchAnalysis {
  home: string;
  away: string;
  neutral: boolean;
  eloHome: number;
  eloAway: number;
  lamHome: number;
  lamAway: number;
  pHome: number;
  pDraw: number;
  pAway: number;
  expectedHomeGoals: number;
  expectedAwayGoals: number;
  overround: number | null;
  outcomes: OutcomeView[];
  topScorelines: Array<{ home: number; away: number; p: number }>;
  dryRun: boolean;
  hasBets: boolean;
  liveState?: LiveState | null;
  scorelineMatrix?: number[][];
}

export interface AnalyzeInput {
  home: string;
  away: string;
  homeOdds?: number | null;
  drawOdds?: number | null;
  awayOdds?: number | null;
  neutral?: boolean;
  eloHome?: number | null;
  eloAway?: number | null;
  liveState?: LiveState | null;
}

function buildOutcomeViews(
  modelPs: Record<Outcome, number>,
  haveOdds: boolean,
  oddsMap: Record<Outcome, number>,
  consensus: Record<Outcome, number>,
  kept: Leg[],
  stakeFracs: number[],
): OutcomeView[] {
  const stakeByOutcome = Object.fromEntries(
    kept.map((leg, idx) => [leg.outcome, stakeFracs[idx]]),
  );

  return OUTCOMES.map((o) => {
    const mp = modelPs[o];
    if (haveOdds) {
      const odds = oddsMap[o];
      const price = 1 / odds;
      const cons = consensus[o];
      const edge = mp - cons;
      const ev = mp / price - 1;
      const frac = stakeByOutcome[o] ?? 0;
      return {
        outcome: o,
        modelP: mp,
        fairOdds: mp > 0 ? 1 / mp : Infinity,
        marketPrice: price,
        marketOdds: odds,
        consensusP: cons,
        edge,
        evPerDollar: ev,
        stakeFraction: frac,
        stakeUsd: frac * DEFAULT_CONFIG.bankroll,
        bet: frac > 0,
      };
    }
    return {
      outcome: o,
      modelP: mp,
      fairOdds: mp > 0 ? 1 / mp : Infinity,
      marketPrice: null,
      marketOdds: null,
      consensusP: null,
      edge: null,
      evPerDollar: null,
      stakeFraction: 0,
      stakeUsd: 0,
      bet: false,
    };
  });
}

function runAnalysisCore(
  input: AnalyzeInput,
  matrix: number[][],
  lamHome: number,
  lamAway: number,
  eh: number,
  ea: number,
): MatchAnalysis {
  const {
    home,
    away,
    homeOdds,
    drawOdds,
    awayOdds,
    neutral = true,
    liveState = null,
  } = input;

  const [pHome, pDraw, pAway] = marginalHda(matrix);
  const modelPs: Record<Outcome, number> = { home: pHome, draw: pDraw, away: pAway };

  const haveOdds =
    validOdds(homeOdds) && validOdds(drawOdds) && validOdds(awayOdds);

  const oddsMap: Record<Outcome, number> = {
    home: homeOdds ?? 0,
    draw: drawOdds ?? 0,
    away: awayOdds ?? 0,
  };

  let consensus: Record<Outcome, number> = { home: 0, draw: 0, away: 0 };
  let overround: number | null = null;
  let legs: Leg[] = [];

  if (haveOdds) {
    const ordered = OUTCOMES.map((o) => oddsMap[o]);
    const devigged = devig(ordered, DEFAULT_CONFIG.devigMethod);
    OUTCOMES.forEach((o, idx) => {
      consensus[o] = devigged[idx];
    });
    overround = ordered.reduce((acc, o) => acc + 1 / o, 0) - 1;
    legs = OUTCOMES.map((o) => ({
      outcome: o,
      p: modelPs[o],
      price: 1 / oddsMap[o],
    }));
  }

  const kept = haveOdds ? filterLegs(legs, DEFAULT_CONFIG.edgeThreshold) : [];
  const stakeFracs = kept.length
    ? jointMatchStakes(kept, DEFAULT_CONFIG.kellyFraction, DEFAULT_CONFIG.kellyCap)
    : [];
  const outcomes = buildOutcomeViews(
    modelPs,
    haveOdds,
    oddsMap,
    consensus,
    kept,
    stakeFracs,
  );

  return {
    home,
    away,
    neutral,
    eloHome: eh,
    eloAway: ea,
    lamHome,
    lamAway,
    pHome,
    pDraw,
    pAway,
    expectedHomeGoals: lamHome,
    expectedAwayGoals: lamAway,
    overround,
    outcomes,
    topScorelines: topScorelines(matrix),
    dryRun: DEFAULT_CONFIG.dryRun,
    hasBets: outcomes.some((o) => o.bet),
    liveState,
    scorelineMatrix: matrix,
  };
}

export function analyzeMatch(input: AnalyzeInput): MatchAnalysis {
  const {
    home,
    away,
    neutral = true,
    eloHome: eloHomeOverride,
    eloAway: eloAwayOverride,
    liveState,
  } = input;

  const eh = eloHomeOverride ?? getRating(home);
  const ea = eloAwayOverride ?? getRating(away);
  const homeAdv = neutral ? 0 : DEFAULT_CONFIG.homeAdvantageElo;
  const [lamHome, lamAway] = lambdasFromElo(eh, ea, homeAdv);

  const matrix =
    liveState != null
      ? conditionalScorelineMatrix(lamHome, lamAway, liveState)
      : scorelineMatrix(lamHome, lamAway);

  return runAnalysisCore(input, matrix, lamHome, lamAway, eh, ea);
}

/** Explicit in-play entry point (same pipeline, conditional matrix). */
export function analyzeLiveMatch(
  input: AnalyzeInput & { liveState: LiveState },
): MatchAnalysis {
  return analyzeMatch(input);
}

File: src/fvmatch/engine.py
"""End-to-end fair-value engine: team strength → model probs → market edge.

This is the glue that makes ``fvmatch`` a usable tool. Given two teams and
(optionally) the three-way match odds, it produces:

* model home/draw/away probabilities from a Dixon-Coles scoreline matrix whose
  goal expectations come from the Elo strength prior,
* the de-vigged market consensus and the per-outcome edge,
* fractional-Kelly stake sizing on the price actually paid, gated by the edge
  threshold.

Everything here is pure/deterministic given its inputs (no network, no DB), so
it runs anywhere and is fully testable.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import cast

import numpy as np
from numpy.typing import NDArray

from fvmatch.config import Settings
from fvmatch.config import settings as default_settings
from fvmatch.edge.devig import Method, devig
from fvmatch.edge.gate import filter_legs
from fvmatch.edge.kelly import Leg, joint_match_stakes
from fvmatch.model.dixon_coles import (
    lambdas_from_elo,
    marginal_hda,
    scoreline_matrix,
)
from fvmatch.model.ratings_prior import get_rating

OUTCOMES = ("home", "draw", "away")


@dataclass(frozen=True)
class OutcomeView:
    outcome: str
    model_p: float
    fair_odds: float
    market_price: float | None  # cost per $1 share = 1/decimal_odds
    market_odds: float | None
    consensus_p: float | None  # de-vigged market probability
    edge: float | None  # model_p - consensus_p
    ev_per_dollar: float | None  # model_p / market_price - 1
    stake_fraction: float
    stake_usd: float
    bet: bool


@dataclass(frozen=True)
class MatchAnalysis:
    home: str
    away: str
    neutral: bool
    elo_home: float
    elo_away: float
    lam_home: float
    lam_away: float
    p_home: float
    p_draw: float
    p_away: float
    expected_home_goals: float
    expected_away_goals: float
    overround: float | None
    outcomes: list[OutcomeView]
    top_scorelines: list[tuple[int, int, float]] = field(default_factory=list)
    dry_run: bool = True

    @property
    def has_bets(self) -> bool:
        return any(o.bet for o in self.outcomes)


def _top_scorelines(
    matrix: NDArray[np.float64], n: int = 6
) -> list[tuple[int, int, float]]:
    flat: list[tuple[int, int, float]] = []
    rows, cols = matrix.shape
    for i in range(rows):
        for j in range(cols):
            flat.append((i, j, float(matrix[i, j])))
    flat.sort(key=lambda t: t[2], reverse=True)
    return flat[:n]


def analyze_match(
    home: str,
    away: str,
    home_odds: float | None = None,
    draw_odds: float | None = None,
    away_odds: float | None = None,
    *,
    neutral: bool = True,
    elo_home: float | None = None,
    elo_away: float | None = None,
    config: Settings | None = None,
    devig_method: Method | None = None,
) -> MatchAnalysis:
    """Run the full fair-value pipeline for a single fixture.

    Args:
        home, away: Team names (looked up in the bundled Elo seed).
        home_odds, draw_odds, away_odds: Decimal odds for the three-way market.
            If omitted, the analysis is model-only (no edge/stakes).
        neutral: True for a neutral venue (no home-advantage Elo bump).
        elo_home, elo_away: Override the looked-up Elo ratings.
        config: Settings instance (defaults to the global ``settings``).
        devig_method: Override the configured de-vig method.

    Returns:
        A populated :class:`MatchAnalysis`.
    """
    cfg = config or default_settings
    eh = elo_home if elo_home is not None else get_rating(home)
    ea = elo_away if elo_away is not None else get_rating(away)
    home_adv = 0.0 if neutral else cfg.home_advantage_elo

    lam_home, lam_away = lambdas_from_elo(
        eh,
        ea,
        home_advantage=home_adv,
        base_goals=cfg.base_goals,
        goal_scale=cfg.elo_goal_scale,
    )
    matrix = scoreline_matrix(
        lam_home, lam_away, rho=cfg.dc_rho, max_goals=cfg.max_goals
    )
    p_home, p_draw, p_away = marginal_hda(matrix)
    model_ps = {"home": p_home, "draw": p_draw, "away": p_away}

    have_odds = (
        home_odds is not None
        and draw_odds is not None
        and away_odds is not None
        and home_odds > 1.0
        and draw_odds > 1.0
        and away_odds > 1.0
    )

    odds_map: dict[str, float] = {}
    consensus: dict[str, float] = {}
    overround: float | None = None
    legs: list[Leg] = []
    if have_odds:
        odds_map = {
            "home": float(home_odds),  # type: ignore[arg-type]
            "draw": float(draw_odds),  # type: ignore[arg-type]
            "away": float(away_odds),  # type: ignore[arg-type]
        }
        ordered = [odds_map[o] for o in OUTCOMES]
        method = cast(Method, devig_method or cfg.devig_method)
        devigged = devig(ordered, method=method)
        consensus = dict(zip(OUTCOMES, devigged, strict=True))
        overround = sum(1.0 / o for o in ordered) - 1.0
        for o in OUTCOMES:
            legs.append(Leg(outcome=o, p=model_ps[o], price=1.0 / odds_map[o]))

    kept = filter_legs(legs, threshold=cfg.edge_threshold) if have_odds else []
    stake_fracs = (
        joint_match_stakes(kept, fraction=cfg.kelly_fraction, cap=cfg.kelly_cap)
        if kept
        else []
    )
    stake_by_outcome = {
        leg.outcome: frac for leg, frac in zip(kept, stake_fracs, strict=True)
    }

    views: list[OutcomeView] = []
    for o in OUTCOMES:
        mp = model_ps[o]
        if have_odds:
            odds = odds_map[o]
            price = 1.0 / odds
            cons = consensus[o]
            edge = mp - cons
            ev = mp / price - 1.0
            frac = stake_by_outcome.get(o, 0.0)
            views.append(
                OutcomeView(
                    outcome=o,
                    model_p=mp,
                    fair_odds=1.0 / mp if mp > 0 else float("inf"),
                    market_price=price,
                    market_odds=odds,
                    consensus_p=cons,
                    edge=edge,
                    ev_per_dollar=ev,
                    stake_fraction=frac,
                    stake_usd=frac * cfg.bankroll,
                    bet=frac > 0.0,
                )
            )
        else:
            views.append(
                OutcomeView(
                    outcome=o,
                    model_p=mp,
                    fair_odds=1.0 / mp if mp > 0 else float("inf"),
                    market_price=None,
                    market_odds=None,
                    consensus_p=None,
                    edge=None,
                    ev_per_dollar=None,
                    stake_fraction=0.0,
                    stake_usd=0.0,
                    bet=False,
                )
            )

    return MatchAnalysis(
        home=home,
        away=away,
        neutral=neutral,
        elo_home=eh,
        elo_away=ea,
        lam_home=lam_home,
        lam_away=lam_away,
        p_home=p_home,
        p_draw=p_draw,
        p_away=p_away,
        expected_home_goals=lam_home,
        expected_away_goals=lam_away,
        overround=overround,
        outcomes=views,
        top_scorelines=_top_scorelines(matrix),
        dry_run=cfg.dry_run,
    )
